packages/SwingSet/src/kernel/kernel.js

```diff
@@ -203,31 +213,27 @@ export default function buildKernel(kernelEndowments) {
     return resultRead;
   }
 
-  async function deliverToVat(vatID, target, msg) {
-    insistMessage(msg);
-    const vat = ephemeral.vats.get(vatID);
-    assert(vat, details`unknown vatID ${vatID}`);
-    kernelKeeper.incStat('dispatches');
-    kernelKeeper.incStat('dispatchDeliver');
-    const kd = harden(['message', target, msg]);
-    const vd = vat.translators.kernelDeliveryToVatDelivery(kd);
-    try {
-      await vat.manager.deliver(vd);
-    } catch (e) {
-      // log so we get a stack trace
-      console.error(`error in kernel.deliver:`, e);
-      throw e;
-    }
-  }
-
-  function getKernelResolveablePromise(kpid) {
+  function getResolveablePromise(kpid, resolvingVatID) {
     insistKernelType('promise', kpid);
+    if (resolvingVatID) {
+      insistVatID(resolvingVatID);
+    }
     const p = kernelKeeper.getKernelPromise(kpid);
     assert(p.state === 'unresolved', details`${kpid} was already resolved`);
-    assert(!p.decider, details`${kpid} is decided by ${p.decider}, not kernel`);
+    assert(
```

Comment:
Taking this apart more thoroughly: in the old world order we had `getKernelResolvablePromise` used when the kernel was the decider and `getResolvablePromise` used when the decider was a specific vat, and each checked in its own way that the decider was who it thought it was.  Having `null`/`undefined` indicate the kernel as the resolver seems fine (and renaming the parameter to `expectedDecider` seems like a clarifying improvement).  The key is that the `!p.decider` test is no longer sufficient for the kernel case, since there are two distinguishable values indicating the kernel and we have to know not only that it's the kernel deciding but that we _expect_ the kernel to be deciding.  So I think if we change the test to `(!p.decider && !expectedDecider) || p.decider === expectedDecider` we should be OK (I think this ends up being the same logic as what you proposed).  I don't think we need to add another parameter to `deliverToError` as the `expectedDecider` param and the `vatToNotNotify` param are _always_ going to be the same.

The other thing that kinda bugs me is that we currently have two different functions named `getResolvablePromise`, one in `kernel.js` and one in `kernelSyscall.js`.  I think the latter can be eliminated but it raises a cross-module linkage ugliness.  We can put it in the bag of stuff that gets passed to `makeKernelSyscallHandler`, but it's a little weird because it doesn't provide any new authority the kernel syscall handler doesn't already have through the other args.

Reply:
aye. Let's use one parameter, `expectedDecider`, and add a comment in the loop that says "we don't notify the decider, they already know". I'd like to stick with the if+else check rather than the boolean reduction (I find it tiring to read expressions like that and verify that they behave the way I intended.. the phrase "DeMorgan's line-noise" comes to mind :).

For `getResolvablePromise`, we could decompose it. The only thing with authority is the kernelKeeper lookup, the rest is just assertions. If it made sense, we could put the assertions in a shared library, which can be imported by each customer without needing to pass anything special through the bag of stuff.

Reply:
I had a co-worker once upon a time who spoke of "drinking too much Captain DeMorgan's Rum".

Reply:
Taking another look at what `getResolveablePromise` does, it just struck me that it would make sense to turn it into a method on `kernelKeeper`.